File: cursor-proxy-debugger/store.go
package proxydebugger

import (
	"sort"
	"strings"
	"sync"
	"time"
)

type exchangeStore struct {
	mu             sync.RWMutex
	maxStoreBytes  int64
	maxExchanges   int
	order          []string // newest first
	exchanges      map[string]*Exchange
	usedBytes      int64
	subscribers    map[chan storeEvent]struct{}
}

func newExchangeStore(maxStoreBytes int64, maxExchanges int) *exchangeStore {
	if maxStoreBytes <= 0 {
		maxStoreBytes = defaultMaxStoreBytes
	}
	return &exchangeStore{
		maxStoreBytes: maxStoreBytes,
		maxExchanges:  maxExchanges,
		exchanges:     make(map[string]*Exchange),
		subscribers:   make(map[chan storeEvent]struct{}),
	}
}

func (store *exchangeStore) create(exchange *Exchange) {
	store.mu.Lock()
	exchange.StoredBytes = estimateExchangeBytes(*exchange)
	store.exchanges[exchange.ID] = exchange
	store.order = append([]string{exchange.ID}, store.order...)
	store.usedBytes += exchange.StoredBytes
	evicted := store.evictLocked()
	store.mu.Unlock()
	store.publish(storeEvent{Type: "created", ID: exchange.ID})
	for _, id := range evicted {
		store.publish(storeEvent{Type: "evicted", ID: id})
	}
}

func (store *exchangeStore) update(id string, apply func(*Exchange)) {
	store.mu.Lock()
	exchange := store.exchanges[id]
	if exchange == nil {
		store.mu.Unlock()
		return
	}
	oldBytes := exchange.StoredBytes
	apply(exchange)
	exchange.StoredBytes = estimateExchangeBytes(*exchange)
	store.usedBytes += exchange.StoredBytes - oldBytes
	if store.usedBytes < 0 {
		store.usedBytes = 0
	}
	evicted := store.evictLocked()
	store.mu.Unlock()
	store.publish(storeEvent{Type: "updated", ID: id})
	for _, eid := range evicted {
		store.publish(storeEvent{Type: "evicted", ID: eid})
	}
}

// evictLocked drops oldest exchanges until under byte/count budgets.
// Caller must hold store.mu.
func (store *exchangeStore) evictLocked() []string {
	var evicted []string
	for len(store.order) > 0 {
		overBytes := store.usedBytes > store.maxStoreBytes
		overCount := store.maxExchanges > 0 && len(store.order) > store.maxExchanges
		if !overBytes && !overCount {
			break
		}
		oldest := store.order[len(store.order)-1]
		store.order = store.order[:len(store.order)-1]
		if exchange := store.exchanges[oldest]; exchange != nil {
			store.usedBytes -= exchange.StoredBytes
			if store.usedBytes < 0 {
				store.usedBytes = 0
			}
			delete(store.exchanges, oldest)
			evicted = append(evicted, oldest)
		} else {
			delete(store.exchanges, oldest)
		}
	}
	return evicted
}

func (store *exchangeStore) summaries() []ExchangeSummary {
	store.mu.RLock()
	defer store.mu.RUnlock()
	result := make([]ExchangeSummary, 0, len(store.order))
	for _, id := range store.order {
		if exchange := store.exchanges[id]; exchange != nil {
			result = append(result, exchange.ExchangeSummary)
		}
	}
	return result
}

func (store *exchangeStore) stats() StoreStats {
	store.mu.RLock()
	defer store.mu.RUnlock()
	return StoreStats{
		Count:         len(store.order),
		UsedBytes:     store.usedBytes,
		MaxStoreBytes: store.maxStoreBytes,
		MaxExchanges:  store.maxExchanges,
	}
}

func (store *exchangeStore) get(id string) (Exchange, bool) {
	store.mu.RLock()
	defer store.mu.RUnlock()
	exchange := store.exchanges[id]
	if exchange == nil {
		return Exchange{}, false
	}
	return cloneExchange(*exchange), true
}

func (store *exchangeStore) query(q ExchangeQuery) (items []Exchange, total int, stats StoreStats) {
	store.mu.RLock()
	defer store.mu.RUnlock()
	stats = StoreStats{
		Count:         len(store.order),
		UsedBytes:     store.usedBytes,
		MaxStoreBytes: store.maxStoreBytes,
		MaxExchanges:  store.maxExchanges,
	}
	if q.Limit <= 0 {
		q.Limit = 50
	}
	if q.Limit > 500 {
		q.Limit = 500
	}
	if q.Offset < 0 {
		q.Offset = 0
	}
	matched := make([]*Exchange, 0, 64)
	for _, id := range store.order {
		exchange := store.exchanges[id]
		if exchange == nil || !exchangeMatches(*exchange, q) {
			continue
		}
		matched = append(matched, exchange)
	}
	total = len(matched)
	if q.Offset >= total {
		return nil, total, stats
	}
	end := q.Offset + q.Limit
	if end > total {
		end = total
	}
	items = make([]Exchange, 0, end-q.Offset)
	for _, exchange := range matched[q.Offset:end] {
		items = append(items, projectExchange(*exchange, q.Include))
	}
	return items, total, stats
}

func exchangeMatches(exchange Exchange, q ExchangeQuery) bool {
	if q.ID != "" && !strings.EqualFold(exchange.ID, q.ID) {
		return false
	}
	if q.Server != "" && !strings.EqualFold(exchange.Server, q.Server) {
		return false
	}
	if q.CaptureSource != "" && !strings.EqualFold(exchange.CaptureSource, q.CaptureSource) {
		return false
	}
	if q.RequestKind != "" && !strings.EqualFold(exchange.RequestKind, q.RequestKind) {
		return false
	}
	if q.ResponseKind != "" && !strings.EqualFold(exchange.ResponseKind, q.ResponseKind) {
		return false
	}
	if q.Method != "" && !strings.EqualFold(exchange.Method, q.Method) {
		return false
	}
	if q.HostContains != "" && !strings.Contains(strings.ToLower(exchange.Host), strings.ToLower(q.HostContains)) {
		return false
	}
	if q.PathContains != "" && !strings.Contains(strings.ToLower(exchange.Path), strings.ToLower(q.PathContains)) {
		return false
	}
	if q.RequestID != "" && !strings.Contains(strings.ToLower(exchange.RequestID), strings.ToLower(q.RequestID)) {
		return false
	}
	if q.Status != 0 && exchange.Status != q.Status {
		return false
	}
	if q.MinReqBytes > 0 && exchange.RequestBytes < q.MinReqBytes {
		return false
	}
	if q.MinRespBytes > 0 && exchange.ResponseBytes < q.MinRespBytes {
		return false
	}
	if !q.Since.IsZero() && exchange.StartedAt.Before(q.Since) {
		return false
	}
	if !q.Until.IsZero() && exchange.StartedAt.After(q.Until) {
		return false
	}
	if q.HasRaw != nil {
		has := exchange.Request.RawHex != "" || exchange.Response.RawHex != ""
		if *q.HasRaw != has {
			return false
		}
	}
	if q.HasDecoded != nil {
		has := exchange.Request.DecodedJSON != "" || exchange.Response.DecodedJSON != "" || len(exchange.Response.Frames) > 0
		if *q.HasDecoded != has {
			return false
		}
	}
	if q.Q != "" {
		needle := strings.ToLower(q.Q)
		haystack := strings.ToLower(strings.Join([]string{
			exchange.ID,
			exchange.Path,
			exchange.URL,
			exchange.Host,
			exchange.RequestID,
			exchange.RequestKind,
			exchange.ResponseKind,
			exchange.Server,
			exchange.CaptureSource,
			exchange.Error,
			exchange.Request.DecodedJSON,
			exchange.Response.DecodedJSON,
		}, "\n"))
		if !strings.Contains(haystack, needle) {
			// also search a few frame json snippets
			found := false
			for _, frame := range exchange.Response.Frames {
				if strings.Contains(strings.ToLower(frame.JSON), needle) || strings.Contains(strings.ToLower(frame.Kind), needle) {
					found = true
					break
				}
			}
			if !found {
				return false
			}
		}
	}
	return true
}

func projectExchange(exchange Exchange, include string) Exchange {
	include = strings.ToLower(strings.TrimSpace(include))
	if include == "" || include == "summary" {
		return Exchange{ExchangeSummary: exchange.ExchangeSummary}
	}
	cloned := cloneExchange(exchange)
	switch include {
	case "full", "all":
		return cloned
	case "decoded":
		cloned.Request.RawHex = ""
		cloned.Response.RawHex = ""
		for i := range cloned.Request.Frames {
			cloned.Request.Frames[i].RawHex = ""
		}
		for i := range cloned.Response.Frames {
			cloned.Response.Frames[i].RawHex = ""
		}
		return cloned
	case "raw":
		cloned.Request.DecodedJSON = ""
		cloned.Response.DecodedJSON = ""
		cloned.Request.Frames = nil
		cloned.Response.Frames = nil
		return cloned
	case "frames":
		cloned.Request.RawHex = ""
		cloned.Response.RawHex = ""
		cloned.Request.DecodedJSON = ""
		cloned.Response.DecodedJSON = ""
		return cloned
	default:
		return Exchange{ExchangeSummary: exchange.ExchangeSummary}
	}
}

func (store *exchangeStore) clear() {
	store.mu.Lock()
	store.order = nil
	store.exchanges = make(map[string]*Exchange)
	store.usedBytes = 0
	store.mu.Unlock()
	store.publish(storeEvent{Type: "cleared"})
}

func (store *exchangeStore) subscribe() (<-chan storeEvent, func()) {
	updates := make(chan storeEvent, 32)
	store.mu.Lock()
	store.subscribers[updates] = struct{}{}
	store.mu.Unlock()
	return updates, func() {
		store.mu.Lock()
		if _, ok := store.subscribers[updates]; ok {
			delete(store.subscribers, updates)
			close(updates)
		}
		store.mu.Unlock()
	}
}

func (store *exchangeStore) publish(event storeEvent) {
	store.mu.RLock()
	defer store.mu.RUnlock()
	for subscriber := range store.subscribers {
		select {
		case subscriber <- event:
		default:
		}
	}
}

func estimateExchangeBytes(exchange Exchange) int64 {
	n := int64(128)
	n += int64(len(exchange.ID) + len(exchange.URL) + len(exchange.Host) + len(exchange.Path) + len(exchange.RequestID))
	n += int64(len(exchange.RequestKind) + len(exchange.ResponseKind) + len(exchange.Error))
	n += estimatePayloadBytes(exchange.Request)
	n += estimatePayloadBytes(exchange.Response)
	return n
}

func estimatePayloadBytes(payload Payload) int64 {
	n := int64(64 + len(payload.ContentType) + len(payload.ContentCodec) + len(payload.DecodedJSON) + len(payload.DecodeError) + len(payload.RawHex))
	for _, header := range payload.Headers {
		n += int64(len(header.Name) + len(header.Value) + 8)
	}
	for _, frame := range payload.Frames {
		n += int64(64 + len(frame.Kind) + len(frame.MessageType) + len(frame.RequestID) + len(frame.JSON) + len(frame.RawHex) + len(frame.Error))
	}
	return n
}

func cloneExchange(exchange Exchange) Exchange {
	exchange.Request = clonePayload(exchange.Request)
	exchange.Response = clonePayload(exchange.Response)
	return exchange
}

func clonePayload(payload Payload) Payload {
	payload.Headers = append([]Header(nil), payload.Headers...)
	payload.Frames = append([]FrameView(nil), payload.Frames...)
	return payload
}

func elapsedMS(startedAt time.Time) int64 {
	if startedAt.IsZero() {
		return 0
	}
	return time.Since(startedAt).Milliseconds()
}

func sortedHeaders(headers map[string][]string) []Header {
	result := make([]Header, 0, len(headers))
	for name, values := range headers {
		value := ""
		for index, item := range values {
			if index > 0 {
				value += ", "
			}
			value += item
		}
		if isSensitiveHeader(name) && value != "" {
			value = "[已隐藏]"
		}
		result = append(result, Header{Name: name, Value: value})
	}
	sort.Slice(result, func(left, right int) bool {
		return result[left].Name < result[right].Name
	})
	return result
}

func isSensitiveHeader(name string) bool {
	switch httpCanonicalLower(name) {
	case "authorization", "cookie", "set-cookie", "proxy-authorization", "x-api-key":
		return true
	default:
		return false
	}
}

func httpCanonicalLower(value string) string {
	buffer := make([]byte, len(value))
	for index := range value {
		character := value[index]
		if character >= 'A' && character <= 'Z' {
			character += 'a' - 'A'
		}
		buffer[index] = character
	}
	return string(buffer)
}

File: server/src/debug/store.rs
//! In-memory capture buffer with byte-budget eviction and live SSE fan-out.
//!
//! Ports the legacy Go `exchangeStore`: newest-last ordering, incremental byte
//! accounting, oldest-first eviction under a byte budget, and a broadcast
//! channel that notifies the web UI of created/updated/evicted/cleared events.
use std::{
    collections::HashMap,
    sync::atomic::{AtomicU64, Ordering},
};

use parking_lot::RwLock;
use serde::Serialize;
use tokio::sync::broadcast;

use super::model::{Exchange, ExchangeState, ExchangeSummary, FrameView, Side, StoreStats};

const MAX_LIST_SUMMARIES: usize = 2000;
/// Publish an SSE update at most every N appended frames to avoid a storm.
const FRAME_PUBLISH_INTERVAL: usize = 64;
const DEFAULT_MAX_FRAMES: usize = 2000;

/// Broadcast event describing a mutation the web UI should react to.
#[derive(Clone, Debug, Serialize)]
pub struct StoreEvent {
    #[serde(rename = "type")]
    pub kind: &'static str,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub id: Option<String>,
}

pub struct ExchangeStore {
    inner: RwLock<Inner>,
    events: broadcast::Sender<StoreEvent>,
    max_store_bytes: i64,
    max_exchanges: usize,
    max_frames: usize,
    next_id: AtomicU64,
}

struct Inner {
    order: Vec<String>,
    exchanges: HashMap<String, Exchange>,
    used_bytes: i64,
}

impl ExchangeStore {
    pub fn new(max_store_bytes: i64, max_exchanges: usize, max_frames: usize) -> Self {
        let (events, _) = broadcast::channel(256);
        Self {
            inner: RwLock::new(Inner {
                order: Vec::new(),
                exchanges: HashMap::new(),
                used_bytes: 0,
            }),
            events,
            max_store_bytes: if max_store_bytes <= 0 {
                200 << 20
            } else {
                max_store_bytes
            },
            max_exchanges,
            max_frames: if max_frames == 0 {
                DEFAULT_MAX_FRAMES
            } else {
                max_frames
            },
            next_id: AtomicU64::new(1),
        }
    }

    pub fn next_id(&self) -> String {
        self.next_id.fetch_add(1, Ordering::Relaxed).to_string()
    }

    pub fn next_prefixed_id(&self, prefix: &str) -> String {
        format!("{}{}", prefix, self.next_id())
    }

    pub fn max_frames(&self) -> usize {
        self.max_frames
    }

    /// Returns the stored content type for one side of an exchange.
    pub fn content_type(&self, id: &str, response_side: bool) -> String {
        let inner = self.inner.read();
        let Some(exchange) = inner.exchanges.get(id) else {
            return String::new();
        };
        if response_side {
            exchange.response.content_type.clone()
        } else {
            exchange.request.content_type.clone()
        }
    }

    pub fn subscribe(&self) -> broadcast::Receiver<StoreEvent> {
        self.events.subscribe()
    }

    pub fn create(&self, mut exchange: Exchange) {
        exchange.recompute_stored_bytes();
        let id = exchange.summary.id.clone();
        let evicted = {
            let mut inner = self.inner.write();
            inner.used_bytes += exchange.summary.stored_bytes;
            inner.order.push(id.clone());
            inner.exchanges.insert(id.clone(), exchange);
            self.evict_locked(&mut inner)
        };
        self.publish("created", Some(id));
        self.publish_evicted(evicted);
    }

    /// Applies `apply` to an existing exchange, keeping byte accounting exact.
    pub fn update(&self, id: &str, apply: impl FnOnce(&mut Exchange)) {
        let evicted = {
            let mut inner = self.inner.write();
            let Some(exchange) = inner.exchanges.get_mut(id) else {
                return;
            };
            let old_bytes = exchange.summary.stored_bytes;
            apply(exchange);
            exchange.recompute_stored_bytes();
            let delta = exchange.summary.stored_bytes - old_bytes;
            inner.used_bytes = (inner.used_bytes + delta).max(0);
            self.evict_locked(&mut inner)
        };
        self.publish("updated", Some(id.to_owned()));
        self.publish_evicted(evicted);
    }

    /// Appends one decoded Connect frame with incremental accounting, throttling
    /// SSE updates and refreshing the summary fields the request list shows.
    pub fn append_frame(&self, id: &str, side: Side, mut frame: FrameView, max_frames: usize) {
        let (should_publish, evicted) = {
            let mut inner = self.inner.write();
            let Some(exchange) = inner.exchanges.get_mut(id) else {
                return;
            };
            let frames_len = exchange.payload_mut(side).frames.len();
            if max_frames > 0 && frames_len >= max_frames {
                return;
            }
            // Successful frames keep only decoded JSON; raw hex is retained only to
            // help debug a decode failure.
            if frame.error.is_empty() {
                frame.raw_hex.clear();
            }
            let delta = frame.estimated_bytes();
            let kind = frame.kind.clone();
            let request_id = frame.request_id.clone();
            let error = frame.error.clone();
            let payload = exchange.payload_mut(side);
            payload.frames.push(frame);
            let frames_len = payload.frames.len();
            if !error.is_empty() {
                payload.decode_error = error;
            }
            exchange.summary.stored_bytes += delta;
            match side {
                Side::Response => {
                    exchange.summary.frame_count = frames_len;
                    if !kind.is_empty() && kind != "end_stream" {
                        exchange.summary.response_kind = kind;
                    }
                }
                Side::Request => {
                    if exchange.summary.frame_count == 0 {
                        exchange.summary.frame_count = frames_len;
                    }
                    if !kind.is_empty() {
                        exchange.summary.request_kind = kind;
                    }
                    if !request_id.is_empty() {
                        exchange.summary.request_id = request_id;
                    }
                }
            }
            inner.used_bytes += delta;
            let should_publish = frames_len % FRAME_PUBLISH_INTERVAL == 0;
            (should_publish, self.evict_locked(&mut inner))
        };
        if should_publish {
            self.publish("updated", Some(id.to_owned()));
        }
        self.publish_evicted(evicted);
    }

    pub fn summaries(&self, limit: usize) -> Vec<ExchangeSummary> {
        let inner = self.inner.read();
        let limit = if limit == 0 || limit > MAX_LIST_SUMMARIES {
            MAX_LIST_SUMMARIES
        } else {
            limit
        };
        let mut result = Vec::with_capacity(limit.min(inner.order.len()));
        // Newest first.
        for id in inner.order.iter().rev() {
            if result.len() >= limit {
                break;
            }
            if let Some(exchange) = inner.exchanges.get(id) {
                result.push(exchange.summary.clone());
            }
        }
        result
    }

    /// Applies the decoded outcome of one side of an exchange to the store.
    ///
    /// Writes raw hex, size, decoded JSON, kind, request id, and backfills
    /// offline/synthetic frames when the streaming path did not populate them.
    pub fn finish_side(
        &self,
        id: &str,
        side: Side,
        path: &str,
        decoded: super::decode::DecodedPayload,
        captured: &[u8],
        truncated: bool,
        read_error: Option<String>,
    ) {
        let captured = captured.to_vec();
        let captured_len = captured.len();
        self.update(id, move |exchange| {
            // Snapshot summary fields before taking the payload borrow.
            let current_kind = match side {
                Side::Response => exchange.summary.response_kind.clone(),
                Side::Request => exchange.summary.request_kind.clone(),
            };
            let request_id = exchange.summary.request_id.clone();

            let mut next_kind;
            let next_request_id;
            {
                let payload = exchange.payload_mut(side);
                payload.size = captured_len as i64;
                payload.raw_hex = hex::encode(&captured);
                payload.raw_truncated = truncated;
                if !decoded.decoded_json.is_empty() {
                    payload.decoded_json = decoded.decoded_json;
                }
                payload.decode_error = decoded.decode_error;
                next_kind = if decoded.kind.is_empty() {
                    current_kind.clone()
                } else {
                    decoded.kind
                };
                next_request_id = if decoded.request_id.is_empty() {
                    request_id
                } else {
                    decoded.request_id
                };
                if payload.frames.is_empty() && !decoded.offline_frames.is_empty() {
                    payload.frames = decoded.offline_frames;
                    if current_kind.is_empty() {
                        next_kind = super::decode::first_non_empty_frame_kind(&payload.frames);
                    }
                }
                if payload.frames.is_empty() && !payload.decoded_json.is_empty() {
                    let frame = super::decode::synthetic_unary_frame(
                        path,
                        &next_kind,
                        &next_request_id,
                        &payload.decoded_json,
                        captured_len,
                    );
                    payload.frames = vec![frame];
                }
            }

            let frame_count = match side {
                Side::Response => exchange.response.frames.len(),
                Side::Request => exchange.request.frames.len(),
            };
            if exchange.summary.frame_count == 0 {
                exchange.summary.frame_count = frame_count;
            }
            match side {
                Side::Request => exchange.summary.request_bytes = captured_len as i64,
                Side::Response => exchange.summary.response_bytes = captured_len as i64,
            }
            match side {
                Side::Response => exchange.summary.response_kind = next_kind,
                Side::Request => exchange.summary.request_kind = next_kind,
            }
            exchange.summary.request_id = next_request_id;
            if side == Side::Response {
                exchange.summary.duration_ms = exchange.elapsed_ms();
                exchange.set_state(ExchangeState::Completed);
            }
            if let Some(error) = read_error {
                exchange.summary.error = error;
                exchange.set_state(ExchangeState::Error);
            }
        });
    }

    pub fn get(&self, id: &str) -> Option<Exchange> {
        self.inner.read().exchanges.get(id).cloned()
    }

    pub fn stats(&self) -> StoreStats {
        let inner = self.inner.read();
        StoreStats {
            count: inner.order.len(),
            used_bytes: inner.used_bytes,
            max_store_bytes: self.max_store_bytes,
            max_exchanges: self.max_exchanges,
        }
    }

    pub fn clear(&self) {
        {
            let mut inner = self.inner.write();
            inner.order.clear();
            inner.exchanges.clear();
            inner.used_bytes = 0;
        }
        self.publish("cleared", None);
    }

    /// Drops oldest exchanges until the byte and count budgets are satisfied.
    fn evict_locked(&self, inner: &mut Inner) -> Vec<String> {
        let mut evicted = Vec::new();
        while !inner.order.is_empty() {
            let over_bytes = inner.used_bytes > self.max_store_bytes;
            let over_count = self.max_exchanges > 0 && inner.order.len() > self.max_exchanges;
            if !over_bytes && !over_count {
                break;
            }
            let oldest = inner.order.remove(0);
            if let Some(exchange) = inner.exchanges.remove(&oldest) {
                inner.used_bytes = (inner.used_bytes - exchange.summary.stored_bytes).max(0);
                evicted.push(oldest);
            }
        }
        evicted
    }

    fn publish(&self, kind: &'static str, id: Option<String>) {
        // Ignore send errors: no subscribers simply means the web UI is closed.
        let _ = self.events.send(StoreEvent { kind, id });
    }

    fn publish_evicted(&self, evicted: Vec<String>) {
        for id in evicted {
            self.publish("evicted", Some(id));
        }
    }
}

/// Projects an exchange for the requested detail level (`include` query param).
pub fn project(mut exchange: Exchange, include: &str) -> Exchange {
    match include.trim().to_ascii_lowercase().as_str() {
        "" | "summary" => Exchange {
            request: Default::default(),
            response: Default::default(),
            summary: exchange.summary,
            started: exchange.started,
        },
        "full" | "all" => exchange,
        "decoded" => {
            strip_raw(&mut exchange);
            exchange
        }
        "raw" => {
            exchange.request.decoded_json.clear();
            exchange.response.decoded_json.clear();
            exchange.request.frames.clear();
            exchange.response.frames.clear();
            exchange
        }
        "frames" => {
            exchange.request.raw_hex.clear();
            exchange.response.raw_hex.clear();
            exchange.request.decoded_json.clear();
            exchange.response.decoded_json.clear();
            exchange
        }
        _ => Exchange {
            request: Default::default(),
            response: Default::default(),
            summary: exchange.summary,
            started: exchange.started,
        },
    }
}

fn strip_raw(exchange: &mut Exchange) {
    exchange.request.raw_hex.clear();
    exchange.response.raw_hex.clear();
    for frame in &mut exchange.request.frames {
        frame.raw_hex.clear();
    }
    for frame in &mut exchange.response.frames {
        frame.raw_hex.clear();
    }
}

pub const fn max_list_summaries() -> usize {
    MAX_LIST_SUMMARIES
}
